main/src/ca/uwaterloo/flix/language/phase/Reducer.scala

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright 2017 Magnus Madsen, Anna Krogh, Patrick Lundvig, Christian Bonde
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationMessage
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Expression
+import ca.uwaterloo.flix.language.ast.{LiftedAst, OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.Validation._
+import ca.uwaterloo.flix.util.Validation
+
+/**
+ * The Reducer phase performs intra-procedural optimizations.
+ *
+ * Specifically,
+ *
+ * - Elimination of dead branches (e.g. if (true) e1 else e2).
+ * - Copy propagation (e.g. let z = w; let y = z; let x = y; x -> w)
+ */
+object Reducer {
+
+  /**
+   * Returns an optimized version of the given AST `root`.
+   */
+  def run(root: OccurrenceAst.Root)(implicit flix: Flix): Validation[LiftedAst.Root, CompilationMessage] = flix.phase("Reducer") {
+    val defs = root.defs.map {
```

Comment:
This looks a bit messy. Try to introduce `visitDef` and use it like it is used in some other phases, e.g. `ClosureConv`

Reply:
Or inside `Inliner`